### ISSUE TITLE
pdf.output instead of save

### ISSUE THREAD

**Issue by djjatoro** [eKoopmans/html2pdf#14]
Hi.

First of all, congratulations about your work. It´s amazing.

I would like to know if with your library is it possible to get the output directly to a var?

For example, with pure jsPDF:

var doc = btoa(pdf.output());

I use it to send as binary to the server, and then attach it to an email.

Otherwise, Do you think is it possible to improve/include, next version?

Sorry my english, please.

Best regards.

**Comment by eKoopmans** [comment]
Hi @djjatoro, thanks! Yes that would be great, and I plan to incorporate that soon. The feature was also requested in #12.

**Comment by dherrero** [comment]
Hi,
Imodified line 214 so that I can open the pdf in a new window:
`if(opt.newWindow){
      pdf.output('dataurlnewwindow');
    }else{
      pdf.save( opt.filename );
    }`
You can try this:
 `if(opt.return){return btoa(pdf.output());}else .....`

And in your JS: 
`var element = document.getElementById('myElementId');`
`var doc = html2pdf(element, {return: true});`

I hope it helps

**Comment by delgah** [comment]
Hi @dherrero, many thanks for this suggestion. I'm trying to call html2pdf with some additional parameters but also I need to get the pdf to a var, so it can be saved to a database. The code is:

------------------------------
var table = document.getElementById("rgnDashboard");
var parameters = {
				margin:			0,
				filename:		'test_jpeg.pdf',
				image:			{ type: 'jpeg', quality: 1 },
                                html2canvas:	{ dpi: 192 },
				jsPDF:			{ unit: 'mm', format: [ width, height], orientation: 'portrait' }
					 };

 html2pdf(table, parameters);
------------------------------
How can I call html2pdf in this way and still get the output as a blob, so it can be saved to a database?

Thanks in advance!

**Comment by eKoopmans** [comment]
Hey guys, I'll have an update soon (hopefully next week) to allow html2pdf to be run in pieces, so that you can get just the PDF object, etc. @delgah for now, if you want to just get the final PDF object as a variable, I've created a [new branch](https://github.com/eKoopmans/html2pdf/tree/pdf-callback). Use the version over there, pass a function in the `pdfCallback` option, and you should be good!

Note, this isn't the approach I'll be using in the upcoming update, it's just a temporary solution. That's why I'm keeping it as a separate branch for now.

**Comment by delgah** [comment]
Many thanks @eKoopmans !! What would be the correct syntax to add the pdfCallback option?
In my case, I have created a test_hd function and I'm calling it in this way, but it's not working:

var parameters = {
		margin:			0,
		//filename:		'test_jpeg.pdf',
               pdfCallback:    {test_hd(doc)},
		image:			{ type: 'jpeg', quality: 1 },
                html2canvas:	{ dpi: 192 },
		jsPDF:			{ unit: 'mm', format: [ width, height], orientation: 'portrait' }
					 };

    html2pdf(table, parameters);

**Comment by eKoopmans** [comment]
Hey @delgah, the `pdfCallback` value should just be a reference to the function, like this:

```js
function myCallback(pdf) {
  // Your function, using the pdf object
}

html2pdf(element, {
  pdfCallback: myCallback,
  // Any other options you want to set...
});
```

**Comment by delgah** [comment]
Hi @eKoopmans, with our custom function we're trying to send the pdf object to an ajax process to finally save the pdf as a blob file into a database table. Do you know if we need to convert first the pdf to a clob or something else before we can send it to the ajax process?

Many thanks

**Comment by delgah** [comment]
Hi @eKoopmans, we found a solution encoding the pdf in base64 and decoding it in a pl/sql process. Regards

**Comment by eKoopmans** [comment]
Great, glad you found a solution!

**Comment by vitormc1** [comment]
@delgah @eKoopmans  can you help me to encode and decode the pdf object? My objetive is send the PDF file by e-mail.

My actual code:

`    function myCallback(pdf) {
            console.log(pdf);
        }

        var element = document.getElementById('tblBoleto');
        html2pdf(element, {
            margin: [10, 5, 0, -60],
            filename: '<?= $dadosboleto["numero_documento"] . '-' . $dadosboleto["sacado_nome"] ?>.pdf',
            image: {type: 'jpeg', quality: 0.98},
            html2canvas: {dpi: 300, letterRendering: true, width: 1300, height: 1000, windowWidth: 1000, windowHeight: 1000},
            jsPDF: {unit: 'mm', format: 'a4', orientation: 'portrait'},
            pdfCallback: myCallback
        });
`

**Comment by eKoopmans** [comment]
Hi @vitormc1, it looks like you received some help already over in #70 and #54. In the future please create a new issue rather than posting in multiple existing issues.

Following up on the other responses, once you have the PDF data with `pdf.toDataURL()`, you can do what you want with it (i.e. email). There's no way to email client-side, so you would need to send the data to some server-side script to send the email for you.

I'll point out that the main purpose of `html2pdf` is to keep things client-side - if your server's getting involved, there are other PDF-creating tools that may be more appropriate, particularly [wkhtmltopdf](https://wkhtmltopdf.org/).

**Comment by eKoopmans** [comment]
Hi, I have just pushed a new branch, [new-api](https://github.com/eKoopmans/html2pdf/tree/new-api), which is ready for testing! [The readme](https://github.com/eKoopmans/html2pdf/tree/new-api#advanced-usage) has all the information on the new API. Please let me know if there are any problems with it - I want to make sure it's stable before merging into master!

For this specific request, you will see that I have added a `.output` method that links into `jsPDF.output`.

**Comment by clintj42** [comment]
I'm getting a "Maximum call stack size exceeded" error when running the new-api branch

**Comment by eKoopmans** [comment]
Interesting, I'll definitely want to fix that! What code were you using/can you show me how to reproduce it?

**Comment by clintj42** [comment]
HTML
`<div id="html-2-pdfwrapper">Here's some text</div>`

Javascript
```
var element = document.getElementById('html-2-pdfwrapper'); 
html2pdf(element);
```

I also tried doing 
`html2pdf(element, options).output('datauristring')`
and got the same error.

**Comment by eKoopmans** [comment]
Thanks. What browser?

**Comment by eKoopmans** [comment]
Hi @clintj42, give it another try, it should be fixed now. There was a recursive definition happening when using `es6-promise` instead of native Promises.

**Comment by clintj42** [comment]
Thanks for looking into this so quickly. Looks like I made it past one error but now I'm getting a "Maximum call stack size exceeded" in a different place. I'm using Chrome and Ionic framework. Here's the stack trace:

```
ShareAttendancePage.html:39 ERROR RangeError: Maximum call stack size exceeded
    at Function.toString (polyfills.js:3)
    at t.setProgress (html2pdf.js:517)
    at t.updateProgress (html2pdf.js:525)
    at then (html2pdf.js:544)
    at t.set$$1 (html2pdf.js:449)
    at new Worker (html2pdf.js:127)
    at t.then (polyfills.js:3)
    at then (html2pdf.js:545)
    at t.set$$1 (html2pdf.js:449)
    at new Worker (html2pdf.js:127)
```

**Comment by eKoopmans** [comment]
Hey @clintj42, thanks. It says `polyfills.js`, do you know what polyfill you are using for Promises?

**Comment by eKoopmans** [comment]
I've tried to generalise the fix for all polyfills, but I'm not that confident about it. See if it works now - and let me know what polyfill you're using either way, thanks!

**Comment by clintj42** [comment]
Still getting the error the polyfill I'm using is "cordova-promise-polyfill": "0.0.2"

**Comment by eKoopmans** [comment]
Hi @gauge421, I haven't been able to reproduce - maybe it's something unique to the Ionic/Cordova framework? I've tried with [`promise-polyfill`](https://github.com/taylorhakes/promise-polyfill) (the basis of [`cordova-promise-polyfill`](https://github.com/rehy/cordova-promise-polyfill)) and with ionic-angular's [`polyfills.js`](https://github.com/fiznool/ionic-angular/blob/master/polyfills/polyfills.js), but both worked fine.

Can you track down what file `polyfills.js` is in your stack/where it comes from? And/or give me a suggestion of how I can reproduce the error on my end? Thanks!

**Comment by TheMadBug** [comment]
Hey, I had the same problem trying to get the new-api branch to work on Ionic.
I was able to fix it by changing the two instances of
```
var selfPromise = (Promise.toString().indexOf('[native code]') === -1) ?
      Worker.convert(Object.assign({}, self), Promise.prototype) : self;
```
to just
`var selfPromise = Worker.convert(Object.assign({}, self), Promise.prototype);`

I even tried removing `es6-promise-plugin` and just use "es6-promise": "4.2.4" but that didn't help. Only the changing of lines seem to.

(I would make a pull request but I really don't know what I'm doing so just thought it best to pass on that information).

**Comment by TheMadBug** [comment]
Well to be accurate, it's not the `[native code]` check that causes the issue, it finds the term `[native code]` at index 30 of the string, and then doesn't do the Worker.convert as a result.

At any rate, thanks for your interest in this. This library has been fantastic for me to generate a PDF as a blob, use ionic-native to save it to a phone's file system, then let people send the resulting PDF around.

To recreate the issue
Install ionic
```
npm install -g cordova ionic
```

Run
```
ionic start helloWorld blank --type=ionic-angular
```

Choose yes when it asks for cordva/ionic-native (not sure if this is needed to re-create the problem but it’s standard use)

Test that this much works with
```
cd helloWorld
ionic serve
```
Quit the serve process and add the dependencies below.

Install new-api and types/node to expose require
(Ionic uses Angular that typically uses typescript, typescript likes to use type definitions for safety if the provided packages are in plain javascript).
```
npm i eKoopmans/html2pdf#new-api --save
npm i @types/node --save
```

Now edit `src/pages/home/home.ts` so it looks like
```
// just below the import statements
const html2pdf = require('html2pdf.js');


// inside the HomePage definition
  public pdf() {
    html2pdf().from('<div>HelloWolrd</div>').save();
  }
```

Edit `src/pages/home/home.html` to add the following inside ion-content
```
<button ion-button (click)="pdf()">PDF</button>
```

Run `ionic serve` again, now you should see a PDF button, when you click it you should get a stack trace of
```
RangeError: Maximum call stack size exceeded
    at JSON.stringify (<anonymous>)
    at new Worker (http://localhost:8100/build/vendor.js:123112:74)
    at t.then (http://localhost:8100/build/polyfills.js:3:22475)
    at then (http://localhost:8100/build/vendor.js:123535:42)
    at t.set$$1 (http://localhost:8100/build/vendor.js:123439:15)
    at new Worker (http://localhost:8100/build/vendor.js:123117:15)
    at t.then (http://localhost:8100/build/polyfills.js:3:22475)
    at then (http://localhost:8100/build/vendor.js:123535:42)
    at t.set$$1 (http://localhost:8100/build/vendor.js:123439:15)
    at new Worker (http://localhost:8100/build/vendor.js:123117:15)
```

**Comment by eKoopmans** [comment]
Thank you again @TheMadBug, I'll investigate as soon as I have a chance!

**Comment by eKoopmans** [comment]
Okay! It's been two months, but I'm back with a fix! I think you'll find the latest commits are now working just fine on Ionic (and more generally, Angular). Thanks again @TheMadBug for the super-clear instructions to test in Ionic, it really helped, once I finally had time to look into it!

In the end, it was just down to Angular doing funny things. It uses its own custom `ZoneAwarePromise` (part of its [`zone.js`](https://github.com/angular/zone.js)) but masks itself as browser-native code. Fortunately, the same fix works for it as for other non-native Promise implementations - it was just a matter of detecting it.

I'm about ready to merge `new-api` down into `master`, as long as no other major issues have cropped up with it. Let me know if there are any further issues!

**Comment by gracielaPosadas** [comment]
Does is any way to use it without es6-promise-plugin ?

**Comment by Monigingir** [comment]
Hi @eKoopmans , 

I am using html2pdf in React. Here is my code 

const element = document.getElementById('test');

    const opt = {
      margin: 1,
      filename: 'myfile.pdf',
      image: { type: 'jpeg', quality: 0.98 },
      html2canvas: { scale: 2 },
      jsPDF: { unit: 'in', format: 'letter', orientation: 'landscape' }
    };

    html2pdf(element, opt);

It downloads a pdf file with all blank pages. Please help.

**Comment by phpmydev** [comment]
Hello eKoopmans!
I'm using straight html2pdf.js in a node app and it works great!
I went through this entire thread just now.
Can I find the code in the documentation for saving the pdf to a database? If so, what do I look for?
Thanks!

**Comment by phpmydev** [comment]
Hi @eKoopmans,
Thanks for your response.
I have already come up with a solution by sending the full html back to the server for insertion into the database as a string.  I then retrieve the html when I need to review the pdf in the same way as first creating it.  I had thought to save the buffer as a blob, but this works as well, if not even better.

**Comment by eKoopmans** [comment]
@phpmydev glad you found a solution that works for you!

**Comment by santosh-hegde** [comment]
`html2pdf().from('<div>HelloWolrd</div>').save();` is working fine in browser. But when I run on my androind using apk file, file is not getting downloaded. `.output`  not working.
How to download file in Android.

**Comment by khairuddeen** [comment]
Hi @eKoopmans ,
i used your code it is easy to create pdf but i want to store file on specific folder and i have a problem in splitting page . the text also splitted 
i need a solution soon as possible
![Capture](https://user-images.githubusercontent.com/60092156/72712844-e5cdfc00-3b91-11ea-8615-ac1ec1b0aeb7.PNG)

**Comment by robatter** [comment]
Hi eKoopmans , how crazy html2pdf!
html2pdf().from('<div>HelloWolrd</div>').save();  The code could download, but I hope it show on the browse directly, how can do, please?

**Comment by kenilvaghasiya** [comment]
I make one website. Here i use html2pdf lib now this web use to android studio.this code not working
